protocol/contracts/templegold/TempleGold.sol

```diff
@@ -0,0 +1,331 @@
+pragma solidity ^0.8.20;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Temple (templegold/TempleGold.sol)
+
+
+import { Origin } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oapp/interfaces/IOAppReceiver.sol";
+import { CommonEventsAndErrors } from "contracts/common/CommonEventsAndErrors.sol";
+import { IOFT, OFTCore } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oft/OFTCore.sol";
+import { ITempleGold } from "contracts/interfaces/templegold/ITempleGold.sol";
+import { IDaiGoldAuction } from "contracts/interfaces/templegold/IDaiGoldAuction.sol";
+import { OFT } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oft/OFT.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ITempleGoldStaking}  from "contracts/interfaces/templegold/ITempleGoldStaking.sol";
+import { MessagingReceipt, MessagingFee } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oapp/OAppSender.sol";
+import { OFTMsgCodec } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oft/libs/OFTMsgCodec.sol";
+import { SendParam, OFTReceipt } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oft/interfaces/IOFT.sol";
+import { TempleMath } from "contracts/common/TempleMath.sol";
+
+/**
+ * @title Temple Gold 
+ * @notice Temple Gold is a non-transferrable ERC20 token with LayerZero integration for cross-chain transfer.
+ * Temple Gold can be only transferred to or from whitelisted addresses. On mint, Temple Gold is distributed between Staking, Auction and Gnosis Safe 
+ * addresses using distribution share percentages set at `DistributionParams`. LayerZero's OFT token standard is modified to allow changing delegates
+ * with the same elevated access from v2. 
+ */
+ contract TempleGold is ITempleGold, OFT {
+    using OFTMsgCodec for bytes;
+    using OFTMsgCodec for bytes32;
+
+    /// @notice These addresses are mutable to allow change/upgrade.
+    /// @notice Staking contract
+    ITempleGoldStaking public override staking;
+    /// @notice Escrow auction contract
+    IDaiGoldAuction public override escrow;
+    /// @notice Multisig gnosis address
+    address public override teamGnosis;
+
+    /// @notice Last block timestamp Temple Gold was minted
+    uint32 public override lastMintTimestamp;
+
+    //// @notice Distribution as a percentage of 100
+    uint256 public constant DISTRIBUTION_MULTIPLIER = 100 ether;
+    /// @notice Minimum percentage of minted Temple Gold to distribute. 1 ether means 1%
+    uint256 public constant MINIMUM_DISTRIBUTION_SHARE = 1 ether;
+    /// @notice 1B max supply
+    uint256 public constant MAX_SUPPLY = 1_000_000_000 ether; // 1B
```

Comment:
Ooo I think I found a good one! What's the bounty?

Because of the LZ integration, when you send tokens to another chain, it burns the tokens. Therefore totalSupply() will be under MAX_SUPPLY and more tokens can be distributed.

Now you could track the actual tokens minted in distribute() separately from the totalSupply...

But an alternative is to just mint the full MAX_SUPPLY at construction (and have the tokens sit here within this contract), and then in distribute you can just transfer instead of mint. Quite nice cause you know how many are left then too

Reply:
bounty is 100 feet rubs ser 😃.  awesome catch!